core/src/main/java/feast/core/model/FeatureSet.java

```diff
@@ -160,6 +152,65 @@ public static FeatureSet fromProto(FeatureSetProto.FeatureSet featureSetProto) {
         featureSetProto.getMeta().getStatus());
   }
 
+  // Updates the existing feature set from a proto.
+  public void updateFromProto(FeatureSetProto.FeatureSet featureSetProto)
+      throws InvalidProtocolBufferException {
+    FeatureSetSpec spec = featureSetProto.getSpec();
+    if (this.toProto().getSpec().equals(spec)) {
+      return;
+    }
+
+    // 1. validate
+    // 1a. check no change to identifiers
+    if (!name.equals(spec.getName())) {
+      throw new IllegalArgumentException(
+          String.format("Given feature set name %s does not match name %s.", spec.getName(), name));
+    }
+    if (!project.getName().equals(spec.getProject())) {
+      throw new IllegalArgumentException(
+          String.format(
+              "You are attempting to change the project of feature set %s from %s to %s. This isn't allowed. Please create a new feature set under the desired project.",
+              spec.getName(), project, spec.getProject()));
+    }
+
+    Set<EntitySpec> existingEntities =
+        entities.stream().map(Entity::toProto).collect(Collectors.toSet());
+
+    // 1b. check no change to entities
+    if (!Sets.newHashSet(spec.getEntitiesList()).equals(existingEntities)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "You are attempting to change the entities of this feature set: Given set of entities \n{%s}\n does not match existing set of entities\n {%s}. This isn't allowed. Please create a new feature set. ",
+              spec.getEntitiesList(), existingEntities));
+    }
+
+    status = FeatureSetStatus.STATUS_PENDING;
+    // 4. Update max age and source.
+    maxAgeSeconds = spec.getMaxAge().getSeconds();
+    source = Source.fromProto(spec.getSource());
+
+    Map<String, FeatureSpec> updatedFeatures =
+        spec.getFeaturesList().stream().collect(Collectors.toMap(FeatureSpec::getName, fs -> fs));
+
+    // 3. Tombstone features that are gone, update features that have changed
+    for (Feature feature : features) {
```

Comment:
Its unclear just by looking at the code where `features` came from. Maybe `this.features`?